### PR TITLE
summary stats to reports

### LINKED ISSUE #355: add summary statistics to calculator, report
#294 added summary statistics to the metrics subpackage. I think we had agreed that the summary statistics should always be calculated for all time series, but it was not implemented because of the big report refactoring.

It's not immediately obvious to me how the data model should change for this. Perhaps we need a `MetricSummaryResult`. Or perhaps the `MetricResult` could be used if `forecast_id: str` became `forecast_id: Union[str, None] = None` and the `__post_init__` method checked that only one attribute is set if `name in metrics.summary._MAP`.

Also not clear what we'd do with `calculate_deterministic_metrics`. Would we adapt that or make a `calculate_summary_metrics`?

Assuming metrics are always calculated, should they always be graphed or only included in the csv file? Should the dashboard have options for actually graphing any of the summary metrics? Should they always be calculated for all categories regardless of the selected report categories?

### PR DESCRIPTION
**Pull request by alorenzo175**
- [x] Closes #355 .
  - [x] I am familiar with the [contributing guidelines](https://solarforecastarbiter-core.readthedocs.io/en/latest/contributing.html).
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes.
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

**Comment by alorenzo175** [comment]
@wholmgren  what do you think so far? This adds a ``MetricResult`` for each processed forecast-obs pair with metrics like `forecast_mean`, `observation_max` etc. with forecast/reference forecasts stats only present for deterministic forecasts. I'll add a table of the totals later and have json/csv downloads available.

**Comment by wholmgren** [comment]
Looks good and is roughly what I had in mind.

File: solarforecastarbiter/datamodel.py [hunk 1]
```diff
@@ -1673,6 +1678,9 @@ class MetricResult(BaseModel):
         UUID of the observation being analyzed.
     aggregate_id: str or None
         UUID of the aggregate being analyzed.
+    is_summary: bool
+        If this metric result is just summary statistics of the
+        observation and forecasts timeseries.
```
Comment (wholmgren) [inline solarforecastarbiter/datamodel.py:1683]:
```suggestion
        If this metric result represents summary statistics of the
        observation and forecasts timeseries.
```

File: solarforecastarbiter/metrics/calculator.py [hunk 2]
```diff
@@ -602,3 +603,123 @@ def _season_from_months(months):
...
+
+
+def _is_deterministic_forecast(proc_fxobs):
+    return (
+        isinstance(proc_fxobs.original.forecast, datamodel.Forecast) and
+        not isinstance(
+            proc_fxobs.original.forecast,
+            (datamodel.ProbabilisticForecast,
+             datamodel.ProbabilisticForecastConstantValue,
+             datamodel.EventForecast)))
```
Comment (wholmgren) [inline solarforecastarbiter/metrics/calculator.py:615]: does `type(proc_fxobs.original.forecast) is datamodel.Forecast` give the same result?

```python
In [1]: class A:
   ...:     pass
   ...:

In [2]: class B(A):
   ...:     pass
   ...:

In [3]: o = B()

In [4]: type(o) is B
Out[4]: True

In [5]: type(o) is A
Out[5]: False

In [6]: a = A()

In [7]: type(a) is A
Out[7]: True

In [8]: type(a) is B
Out[8]: False
```

File: solarforecastarbiter/reports/template.py [hunk 3]
```diff
@@ -39,13 +39,48 @@ def build_metrics_json(report):
...
+    report: :py:class:`solarforecastarbiter.datamodel.Report`
+
+    Returns
+    -------
+    str
+        The json representing the summary statistics. Will be a string
+        representing an empty json array if the report does not have a
+        computed raw_report.
+
+
```
Comment (wholmgren) [inline solarforecastarbiter/reports/template.py:64]:
```suggestion
```

File: solarforecastarbiter/reports/template.py [hunk 4]
```diff
@@ -39,13 +39,48 @@ def build_metrics_json(report):
...
+    Raises
+    ------
+    ValueError
+        If no MetricResults the 'is_summary == True'
```
Comment (wholmgren) [inline solarforecastarbiter/reports/template.py:68]: ?

File: solarforecastarbiter/reports/templates/html/macros.j2 [hunk 5]
```diff
@@ -50,6 +50,59 @@ forecasts/single/{{ forecast.forecast_id }}
...
+<table class="table table-striped table-bordered summary-stats-table-vert" style="width:100%;">
+  <caption style="caption-side:top; text-align: left">
+    Table of {{ human_categories[category] }} data summary statistics
+  </caption>
+  <thead>
+    <tr class="header">
+      <th scope="col" style="text-align: left;">Aligned Pair</th>
+      <th scope="col" colspan="{{ span }}" style="text-align: left;">Observation</th>
+      <th scope="col" colspan="{{ span }}" style="text-align: left;">Forecast</th>
+      <th scope="col" colspan="{{ span }}" style="text-align: left;">Reference Forecast</th>
```
Comment (wholmgren) [inline solarforecastarbiter/reports/templates/html/macros.j2:65]: I think align center would be more consistent with the other tables, especially the "Table of data alignment parameters". Not totally opposed to align left.

**Review by wholmgren — APPROVED** [review]
Locally rendered report looks good and download links work.

File: solarforecastarbiter/metrics/calculator.py (hunk 2 above)
Comment (alorenzo175) [reply solarforecastarbiter/metrics/calculator.py:615]: yes, although I remember some discussion about using type vs isinstance somewhere else. In this case we probably don't want inheritance to work by default so type is more appropriate

**Pull request merged**